File: src/_cpp/TowerDefenseBase/game/unit/UnitWithFadeEffects.cpp
//
//  IslandDefense
//
//  Created by Vladimir Tolmachev on 27.09.14.
//
//
#include "UnitWithFadeEffects.h"
#include "consts.h"
#include "GameGS.h"
#include "tower.h"
#include "UserData.h"
NS_CC_BEGIN


namespace
{
	const float duration = 0.5f;
};

UnitWithFadeEffects::UnitWithFadeEffects()
: _time_ice( 0 )
, _time_fire( 0 )
, _time_electro( 0 )
, _fire( false )
, _ice( false )
, _electro( false )
{}

UnitWithFadeEffects::~UnitWithFadeEffects()
{}

bool UnitWithFadeEffects::init( const std::string & path, const std::string & xmlFile )
{
	do
	{
		CC_BREAK_IF( !Unit::init( path, xmlFile ) );

		setCascadeColorEnabled( true );
		return true;
	}
	while( false );
	return false;
}

void UnitWithFadeEffects::load( const pugi::xml_node & root )
{
	Unit::load( root );

}

bool UnitWithFadeEffects::setProperty( const std::string & name, const std::string & value )
{
	if( name == "color_fire" )
		_color_fire = strToColor3B( value.empty() ? "FF0000" : value );
	else if( name == "color_ice" )
		_color_ice = strToColor3B( value.empty() ? "00FFFF" : value );
	else if( name == "color_electro" )
		_color_electro = strToColor3B( value.empty() ? "FFFF00" : value );
	else
		return Unit::setProperty(name, value);
	return true;
		
}


void UnitWithFadeEffects::update( float dt )
{
	Unit::update( dt );

	Color3B color = Color3B::WHITE;
	
	mlEffect & effect = getEffect( );
	if( effect.negative.fire.damageTime > 0 ||
	    effect.negative.ice.damageTime > 0 ||
	    effect.negative.electro.damageTime > 0
	   )
	{
		bool F = effect.negative.referentialExtendedDamageFire > 0 &&
			effect.positive.fireResist == 1;
		bool I = effect.negative.referentialExtendedDamageIce > 0 &&
			effect.positive.iceResist == 1;
		bool E = effect.negative.referentialExtendedDamageElectro > 0 &&
			effect.positive.electroResist == 1;
		int count = F + I + E;
		
		if( count == 1 )
		{
			_fire = F;
			_ice = I;
			_electro = E;
		}
		else if ( count > 1 )
		{
			if( F && _time_fire < duration * 2 ) { _fire = true; _ice = false; _electro = false; }
			else if( F && _fire && (_time_fire > duration * 2) ) _fire = false;
			if( I && !_fire && _time_ice < duration * 2 ) { _fire = false; _ice = true; _electro = false; }
			else if( I && !_fire && _ice && (_time_ice > duration * 2) ) _ice = false;
			if( E && !_fire && !_ice && _time_electro < duration * 2 ) { _fire = false; _ice = false; _electro = true; }
			else if( E && !_fire && !_ice && _electro && (_time_electro > duration * 2) ) _electro = false;

			assert( (_fire + _ice + _electro) == 1 );
		}
		else
		{
			_fire = false;
			_ice = false;
			_electro = false;
		}


		if( _fire )
		{
			if( _time_fire > duration * 2 ) _time_fire -= duration * 2;
			_time_fire += dt;			
			float t = _time_fire / duration;
			if( t > 1 ) t = 2 - t;

			Vec3 a( 1, 1, 1 );
			Vec3 b = Vec3( _color_fire.r, _color_fire.g, _color_fire.b ) / 255;
			Vec3 c = (a + (b - a) * t) * 255;
			color = Color3B( c.x, c.y, c.z );
		}
		else if( _ice )
		{
			_time_electro += dt;
			if( _time_electro > duration * 2 ) _time_electro -= duration * 2;
			float t = _time_electro / duration;
			if( t > 1 ) t = 2 - t;

			Vec3 a( 1, 1, 1 );
			Vec3 b = Vec3( _color_ice.r, _color_ice.g, _color_ice.b ) / 255;
			Vec3 c = (a + (b - a) * t) * 255;
			color = Color3B( c.x, c.y, c.z );
		}
		else if( _electro )
		{
			_time_electro += dt;
			if( _time_electro > duration * 2 ) _time_electro -= duration * 2;
			float t = _time_electro / duration;
			if( t > 1 ) t = 2 - t;

			Vec3 a( 1, 1, 1 );
			Vec3 b = Vec3( _color_electro.r, _color_electro.g, _color_electro.b ) / 255;
			Vec3 c = (a + (b - a) * t) * 255;
			color = Color3B( c.x, c.y, c.z );
		}

	}

	setColor( color );
}

NS_CC_END

File: src/_eu/game/unit/UnitWithFadeEffects.js
/******************************************************************************
 * Copyright 2014-2016 Vladimir Tolmachev
 * Copyright 2016 Visionarity AG
 * Vladimir Tolmachev and Visionarity AG have unlimited commercial
 * licenses for commercial use and customization
 *
 * Author: Vladimir Tolmachev (tolm_vl@hotmail.com)
 * Ported C++ to Javascript: Visionarity AG / Vladimir Tolmachev
 * Project: Island Defense (JS)
 * If you received the code not from the author, please contact us
 ******************************************************************************/

//Define namespace
var EU = EU || {};

EU.UnitWithFadeEffects = EU.Unit.extend({

    //bool init( const std::string & path, const std::string & xmlFile = "ini.xml" );
    _time_ice : 0.0,
    _time_fire : 0.0,
    _time_electro : 0.0,
    _fire : false,
    _ice : false,
    _electro : false,
    /** Color3B* */ _color_fire : null,
    /** Color3B* */ _color_ice : null,
    /** Color3B* */ _color_electro : null,

    duration : 0.5,
    
    ctor: function(path, xmlFile )
    {
        this._super();
        if ( EU.Unit.prototype.init_str_str.call(this, path, xmlFile ) ) return false;
    
        this.setCascadeColorEnabled( true );
        return true;
    },
    /**@type {Element} root */
    load : function(root )
    {
        EU.Unit.prototype.load.call( this, root );
    },
    setProperty: function( name, value )
    {
        if( name == "color_fire" )
            this._color_fire = cc.hexToColor( EU.xmlLoader.stringIsEmpty(value) ? "FF0000" : value );
        else if( name == "color_ice" )
            this._color_ice = cc.hexToColor( EU.xmlLoader.stringIsEmpty(value) ? "00FFFF" : value );
        else if( name == "color_electro" )
            this._color_electro = cc.hexToColor( EU.xmlLoader.stringIsEmpty(value) ? "FFFF00" : value );
        else
            return EU.Unit.prototype.setProperty.call(this, name, value);
        return true;
    },
    update: function( dt )
    {
        EU.Unit.prototype.update.call(this, dt );
        var Vec3 = cc.math.Vec3;
    
        var color = cc.WHITE;
    
        var effect = this.getEffect( );
        if( effect.negative.fire.damageTime > 0 ||
            effect.negative.ice.damageTime > 0 ||
            effect.negative.electro.damageTime > 0
        )
        {
            var F = effect.negative.referentialExtendedDamageFire > 0 && effect.positive.fireResist == 1;
            var I = effect.negative.referentialExtendedDamageIce > 0 && effect.positive.iceResist == 1;
            var E = effect.negative.referentialExtendedDamageElectro > 0 && effect.positive.electroResist == 1;
            var count = F + I + E;
    
            if( count == 1 )
            {
                this._fire = F;
                this._ice = I;
                this._electro = E;
            }
            else if ( count > 1 )
            {
                if( F && this._time_fire < duration * 2 ) { this._fire = true; this._ice = false; this._electro = false; }
                else if( F && this._fire && (this._time_fire > duration * 2) ) this._fire = false;
                if( I && !this._fire && this._time_ice < duration * 2 ) { this._fire = false; this._ice = true; this._electro = false; }
                else if( I && !this._fire && this._ice && (this._time_ice > duration * 2) ) this._ice = false;
                if( E && !this._fire && !this._ice && this._time_electro < duration * 2 ) { this._fire = false; this._ice = false; this._electro = true; }
                else if( E && !this._fire && !this._ice && this._electro && (this._time_electro > duration * 2) ) this._electro = false;
    
                EU.assert( (this._fire + this._ice + this._electro) == 1 );
            }
            else
            {
                this._fire = false;
                this._ice = false;
                this._electro = false;
            }

            if( this._fire )
            {
                if( this._time_fire > duration * 2 ) this._time_fire -= duration * 2;
                this._time_fire += dt;
                var t = this._time_fire / duration;
                if( t > 1 ) t = 2 - t;
    
                var a = new Vec3( 1, 1, 1 );
                var b = (new Vec3( this._color_fire.r, this._color_fire.g, this._color_fire.b)).scale(1.0 / 255) ;
                var c = (a.add(b.subtract(a)).scale(t)).scale(255);
                color = new cc.Color( c.x, c.y, c.z );
            }
            else if( this._ice )
            {
                this._time_electro += dt;
                if( this._time_electro > duration * 2 ) this._time_electro -= duration * 2;
                var t = this._time_electro / duration;
                if( t > 1 ) t = 2 - t;

                var a = new Vec3( 1, 1, 1 );
                var b = new Vec3( this._color_ice.r, this._color_ice.g, this._color_ice.b).scale(1.0 / 255);
                var c = (a.add(b.subtract(a)).scale(t)).scale(255);
                color = new cc.Color( c.x, c.y, c.z );
            }
            else if( this._electro )
            {
                this._time_electro += dt;
                if( this._time_electro > duration * 2 ) this._time_electro -= duration * 2;
                var t = this._time_electro / duration;
                if( t > 1 ) t = 2 - t;

                var a = new Vec3( 1, 1, 1 );
                var b = new Vec3( this._color_electro.r, this._color_electro.g, this._color_electro.b ).scale(1.0 / 255);
                var c = (a.add(b.subtract(a)).scale(t)).scale(255);
                color = new cc.Color( c.x, c.y, c.z );
            }
    
        }
    
        this.setColor( color );
    }

});